0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+
+What we will instead have to do is change ``ppr`` to have the following type:::
+
+    class Outputable a where
+        ppr :: a -> SDoc b
+
+By being polymorphic in the annotation type, we can at least recover the
+ability to pretty-print the same values in documents that use different
+annotation types. We will however have to arrange for the annotations to
+be produced outside of those ``Outputable`` instances. The ``OutputableBndr``
+class would have to be updated in a similar manner:::
+
+  class Outputable a => OutputableBndr a where
+    pprBndr :: BindingSite -> a -> SDoc b
+    pprBndr _b x = ppr x
+
+    pprPrefixOcc, pprInfixOcc :: a -> SDoc b
+
+    bndrIsJoin_maybe :: a -> Maybe Int
+    bndrIsJoin_maybe _ = Nothing
+
+A good chunk of the work required for implementing this proposal will most
+likely consist in adapting a lot of code in GHC that takes or returns
+``SDoc`` values, and decide whether the annotation type should be
+``Void``, ``ErrAnnotation`` or left polymorphic. Any implementation of this
+proposal should also make sure that the current rendering of error messages
+and IR dumps is not affected, in particular by the changes to the
+pretty-printing infrastructure that are going to be required to perform
+accurate layout computations in the presence of annotations.
+
+Alternatives
+------------
+
+The design for annotated documents as described in this proposal is based
+on the approach used by the *wl-pprint-extras* library, and lets us stick
+annotations at the leaves of our "document trees", and is sometimes referred
+to as the "*pointed* annotations" approach. An alternative design, used for
+example in the Idris compiler, conists in introducing *scoped* annotations:::
+
+    data Doc a
+      = ...
+      | Ann a (Doc a)
+
+where the annotation wraps a sub-document, attaching non-textual information
+to it. This approach has a few drawbacks in our case:
+
+* We want to delay rendering, and the two most obvious ways to use this design
+  would be to attach an annotation to either an empty document to emulate
+  our pointed annotations approach, or to a textual version of the annotation.
+  We are not guaranteed that this text is the one that's going to be used
+  further down the road when reporting errors, since one of the applications of
+  this proposal is to allow tooling authors to customize how some error
+  message entities are rendered.
+
+* This variant of ``Doc`` does not seem to come with lawful ``Applicative``
+  and ``Monad`` instances, which provide a familiar and rich toolbox for
+  introducing, transforming and eliminating annotations.
+
+Unresolved Questions
+--------------------
+
+The only aspect of the implementation that is not crystal clear at this point
+is the handling of annotation nodes in a few key functions from
+``compiler/utils/Pretty.hs``. Fortunately, any implementation that does not
+preserve the current layout bit for bit will quite likely be caught by the
```

Comment:
Is this actually true? I thought the testsuite normally ignores whitespace changes.

Reply:
I'm thinking that would be significant enough to cause more than that, perhaps shifting some contents to the next line. Worst case, I could tweak the driver, to make tests sensible to whitespace changes, temporarily. But you have a point, I should make my statement a bit more precise/honest.

Reply:
It indeed would be nice to be able to ask the testsuite for whitespace-sensitivity. Not just for this work, but in general, in case some other patch proves to be whitespace-sensitive in the future.